core/src/main/java/org/apache/iceberg/avro/AvroSchemaWithTypeVisitor.java

```diff
@@ -79,11 +79,29 @@ private static <T> T visitRecord(Types.StructType struct, Schema record, AvroSch
   private static <T> T visitUnion(Type type, Schema union, AvroSchemaWithTypeVisitor<T> visitor) {
     List<Schema> types = union.getTypes();
     List<T> options = Lists.newArrayListWithExpectedSize(types.size());
-    for (Schema branch : types) {
-      if (branch.getType() == Schema.Type.NULL) {
-        options.add(visit((Type) null, branch, visitor));
-      } else {
-        options.add(visit(type, branch, visitor));
+
+    // simple union case
+    if (AvroSchemaUtil.isOptionSchema(union)) {
+      for (Schema branch : types) {
+        if (branch.getType() == Schema.Type.NULL) {
+          options.add(visit((Type) null, branch, visitor));
+        } else {
+          options.add(visit(type, branch, visitor));
+        }
+      }
+    } else { // complex union case
+      Preconditions.checkArgument(type instanceof Types.StructType,
+          "Cannot visit invalid Iceberg type: %s for Avro complex union type: %s", type, union);
```

Comment:
I think this should also check whether the schema with type visitor has the `tag` field. There's no guarantee that it does.

Reply:
How about aligning by the type? `field_i` of type `x` aligns to the option of type `x`, regardless of the order? Else, we can mandate that the struct is in the same order as the options order (and the types match), and throw an exception here if not. I think both require recursively visiting the types to check for equality, but should be doable. The latter is kind of implemented here already, but I guess it will fail when trying to match the children as opposed to failing when trying to match the union itself.

Reply:
This PR is relevant too (still internal but will be brought upstream soon): https://github.com/linkedin/iceberg/pull/108

Reply:
> How about aligning by the type?

I think that's what we will need to do at some point, but this visitor assumes that both schemas have field IDs. I think for this, the right way to handle it is to get the field ID from the union type. It would mean rewriting the Avro schema ahead of time to look like this:

```json
[
  "null",
  {"type": "int", "field-id": 34},
  {"type": "string", "field-id": 35}
]
```

That's why I'm wondering about how to attach the field IDs in the name mapping. In the name mapping, we could allow a nested level to represent the union. Names in that level could be types rather than names, so the mapping to produce the union above would be `[ { "field-id": 34, "names": ["int"] }, { "field-id": 35, "names": ["string"] } ]`. That works for simple types. For record, map, and array types we can use the simple type name as well, `"record"` or `"map"` or `"array"`. That would support any union with just one option of each nested type. If you had more than one map in the union, it would fail. I think that's a reasonable starting place, though.

@wmoustafa, what do you think?

Reply:
What about the second option, where we expect them to be in the same order? (see this [PR](https://github.com/linkedin/iceberg/pull/108) to support missing fields in the case of projection pruning)? This approach will also make sure that the deep types also match.
For the above suggestion, I am a bit worried about using only the top level types since it will fail in unexpected places and could lead to cryptic error messages if things go wrong. Also, can we list the whole type hierarchy instead? 

That said, does it need to be retrofit into name mapping? I feel we could implement it without extending the name mapping as long as we have functions to deeply compare types.

I think as a starting point, we could assume they must align in terms of order (but do not necessarily be the same, as we could skip some in the struct), and later we can implement deep type comparisons, which will relax the ordering expectation (which is also forward compatible with the type-based check).

Reply:
> I am a bit worried about using only the top level types since it will fail in unexpected places and could lead to cryptic error messages if things go wrong

We can fail gracefully. For example, if there are two records, we throw an exception in name mapping that multiple records aren't supported. If you think this is a common case, we can go further to find a solution for arbitrary nested types. That isn't too hard, actually. We could do it based on the child field set, which would have to match or at least have some overlap.

I don't think that doing this by order is a good idea. That could easily lead to worse cases where we're returning the wrong data.

Reply:
> I don't think that doing this by order is a good idea. That could easily lead to worse cases where we're returning the wrong data.

Is the concern because of an out-of-order schema (e.g., the reader schema or the expected schema)? @yiqiangin has tried both cases and an out of order reader schema throws an exception and an out of order expected schema still returns correct results (both after applying [this patch](https://github.com/linkedin/iceberg/pull/108) to address missing fields/projection pruning, so we may need to take that into account).

Reply:
The problem is that there can be identical branches in a union and order-based resolution could do this incorrectly. The name mapping approach allows this to be entirely by ID here, and the name mapping could do deeper validation because it has child field IDs for all nested types.